jbmc/src/janalyzer/CMakeLists.txt

```diff
@@ -30,6 +30,6 @@ install(TARGETS janalyzer DESTINATION ${CMAKE_INSTALL_BINDIR})
 install(CODE "execute_process( \
   COMMAND ${CMAKE_COMMAND} -E create_symlink \
   cbmc.1 \
-  ${CMAKE_INSTALL_FULL_MANDIR}/man1/janalyzer.1 \
+  $ENV{DESTDIR}${CMAKE_INSTALL_FULL_MANDIR}/man1/janalyzer.1 \
```

Comment:
So I’ve done some investigating and I believe the only thing we really _need_ to change here would be simply to escape the `$ENV` to delay variable expansion to install time, and I’d also add a slash so it doesn’t break if we ever use a base directory that’s not absolute (not the case in your examples because you use full paths, but we want to generalise this as much as possible I’d think).

```suggestion
  \$ENV{DESTDIR}/${CMAKE_INSTALL_FULL_MANDIR}/man1/janalyzer.1 \
```
Because two leading slashes has an [implementation defined meaning](https://unix.stackexchange.com/questions/1910/how-does-linux-handle-multiple-consecutive-path-separators-home-username) and [CMake recommends against the use of DESTDIR on Windows](https://cmake.org/cmake/help/v3.18/envvar/DESTDIR.html)(primarily because windows absolute paths start with a drive letter, which doesn’t make sense in the middle of a path), I’d actually recommend something along the lines of:

```
install(CODE
  "set(CPROVER_INSTALL_MAN_SYMLINK_DESTINATION \"${CMAKE_INSTALL_FULL_MANDIR}\")\n\
   if(DEFINED ENV{DESTDIR} AND NOT WIN32) \n \
     string(PREPEND CPROVER_INSTALL_MAN_SYMLINK_DESTINATION \"\$ENV{DESTDIR}\") \n \
   endif() \n \
   execute_process(COMMAND ${CMAKE_COMMAND} -E create_symlink cbmc.1 \"\${CPROVER_INSTALL_MAN_SYMLINK_DESTINATION}/cbmc.1\")")
```

And since the logic for this is basically always going to be the same and messing around with sometimes-quoted-sometimes-not strings is super error prone, we’re better off extracting this into a function:

```
# We use this function to install symlinks to their proper places -- in
# particular, it is important we take build-time DESTDIR into account
# Usage:
#   cprover_install_symlink(
#     TARGET what-you-want-to-link-to
#     DESTINATION where-the-link-should-go)
# Note that on the CMake documentation (https://cmake.org/cmake/help/v3.18/envvar/DESTDIR.html)
# it notes that DESTDIR should not be used on windows, so we don’t take it into consideration
# if WIN32 is set.
function(cprover_install_symlink)
  cmake_parse_arguments(CPROVER_INSTALL_SYMLINK "" "TARGET;DESTINATION" "" ${ARGN})
  install(CODE
    "set(CPROVER_INSTALL_SYMLINK_DESTINATION \"${CPROVER_INSTALL_SYMLINK_DESTINATION}\") \n \
     if(DEFINED ENV{DESTDIR} AND NOT WIN32) \n \
       string(PREPEND CPROVER_INSTALL_SYMLINK_DESTINATION \"\$ENV{DESTDIR}/\") \n \
     endif() \n \
     execute_process( \
       COMMAND \"${CMAKE_COMMAND}\" \
         -E create_symlink \
         \"${CPROVER_INSTALL_SYMLINK_TARGET}\" \ 
         \"\${CPROVER_INSTALL_SYMLINK_DESTINATION}\")")
endfunction()
```

which for this case you’d call as

```
cprover_install_symlink(
  TARGET cbmc.1
  DESTINATION "${CMAKE_INSTALL_FULL_MANDIR}/man1/janalyzer.1")
```

Unfortunately CMake currently does not have a builtin function for ergonomic path concatenation: https://gitlab.kitware.com/cmake/cmake/-/issues/19568#note_704230, but this way we should have double slashes only in the middle of paths where I don’t think they’re likely to cause problems.

If you could add this function and use it throughout the PR I think that’d address the concerns we both raised. LMK if you have any questions or otherwise need help with anything.

Reply:
I’ve also been thinking this should probably issue a warning if the `DESTDIR` variable is defined on windows, as again, it won’t actually work there.